### PR TITLE
xds: Use a grpcsync.Event to represent close in CDS LB policy.

### PR DESCRIPTION
**Pull request by easwars**

### CODE REVIEW

**Comment by easwars** [comment]
The diff in `run()` shows up much bigger than it actually is. Only the handling of the close event is moved to a separate case on a new `select`, the others are actually unchanged.

File: xds/internal/balancer/cdsbalancer/cdsbalancer.go [hunk 1]
```diff
@@ -178,78 +170,80 @@ type cdsBalancer struct {
...
 			}
-			ccState := balancer.ClientConnState{
-				ResolverState:  resolver.State{Attributes: attributes.New(xdsinternal.XDSClientID, b.client)},
-				BalancerConfig: lbCfg,
-			}
-			if err := b.edsLB.UpdateClientConnState(ccState); err != nil {
-				b.logger.Errorf("xds: edsBalancer.UpdateClientConnState(%+v) returned error: %v", ccState, err)
-			}
-		case *closeUpdate:
+		case <-b.closed.Done():
```
Comment (menghanl) [inline xds/internal/balancer/cdsbalancer/cdsbalancer.go:246]: There's a 50/50 chance to keep processing the update after close. What makes things worse is, `b.updateCh` is a queue, so there can be several items in the queue, or added after close.

Check `b.closed` in `case u := <-b.updateCh.Get():`?

**Review by menghanl — CHANGES_REQUESTED** [review]

File: xds/internal/balancer/cdsbalancer/cdsbalancer.go (hunk 1 above)
Comment (easwars) [reply xds/internal/balancer/cdsbalancer/cdsbalancer.go:246]: We do check whether `closed` has fired before pushing stuff in to `b.updateCh`. So, items getting added to it after close doesn't seem to be possibility to me. And with this approach, we might close early even if there were stuff added to the `b.updateCh` before close was called, which I think is still fine. What do you think?


I could add another check for `b.closed` in `case u := <-b.updateCh.Get():`? But still not convinced that it is required ...

Comment (menghanl) [reply xds/internal/balancer/cdsbalancer/cdsbalancer.go:246]: Yeah, that makes sense. I missed that we won't push to the queue after close.

**Review by menghanl — APPROVED** [review]

**Pull request merged**